### PR TITLE
feat: Better errors

### PR DESCRIPTION
**Pull request by joshuajbouw**
This PR came out out of this issue outlined here: https://github.com/aurora-is-near/aurora-engine/pull/48#discussion_r624676151.

Do note, that this PR may look intimidating with amount of files changed + lines, but most of it is refactoring.

This is in general just a much better error structure. I really hated how the underlying EVM for some bizarre reason mixed together BOTH result types. It just can be awkward to work with in general. That is fixed with this PR.

As a bonus, all error returns are now all consistent too as `ERR_`.

Point of discussion: As far as I'm aware, Reverts shouldn't be considered an "ERR" as they are part of everyday EVM life? As in, the contract executes correctly, but due to some circumstance, it'll back out without actually applying changes.

Closes: https://github.com/aurora-is-near/aurora-engine/issues/33

### CODE REVIEW

File: src/engine.rs [hunk 1]
```diff
@@ -144,14 +230,12 @@ impl Engine {
...
-    fn check_increase_balance(address: &Address, amount: &U256) -> Result<U256, ExitError> {
+    fn check_increase_balance(address: &Address, amount: &U256) -> Result<U256, EngineError> {
         let balance = Self::get_balance(address);
         if let Some(new_balance) = balance.checked_add(*amount) {
             Ok(new_balance)
         } else {
-            Err(ExitError::Other(Borrowed(
-                "balance is too high, can not increase",
-            )))
+            Err(ExitError::Other(Borrowed("balance is too high, can not increase")).into())
```
Comment (artob) [inline src/engine.rs:238]: Missed this one (not an error code)

File: src/engine.rs [hunk 2]
```diff
@@ -162,14 +246,12 @@ impl Engine {
...
-    fn check_decrease_balance(address: &Address, amount: &U256) -> Result<U256, ExitError> {
+    fn check_decrease_balance(address: &Address, amount: &U256) -> Result<U256, EngineError> {
         let balance = Self::get_balance(address);
         if let Some(new_balance) = balance.checked_sub(*amount) {
             Ok(new_balance)
         } else {
-            Err(ExitError::Other(Borrowed(
-                "balance is too low, can not decrease",
-            )))
+            Err(ExitError::Other(Borrowed("balance is too low, can not decrease")).into())
```
Comment (artob) [inline src/engine.rs:254]: Missed this one (not an error code)

File: src/engine.rs [hunk 3]
```diff
@@ -1,14 +1,100 @@
...
+            EvmError(ExitError::InvalidRange) => "ERR_INVALID_RANGE",
+            EvmError(ExitError::DesignatedInvalid) => "ERR_DESIGNATED_INVALID",
+            EvmError(ExitError::CallTooDeep) => "ERR_CALL_TOO_DEEP",
+            EvmError(ExitError::CreateCollision) => "ERR_CREATE_COLLISION",
+            EvmError(ExitError::CreateContractLimit) => "ERR_CREATE_CONTRACT_LIMIT",
+            EvmError(ExitError::OutOfOffset) => "ERR_OUT_OF_OFFSET",
+            EvmError(ExitError::OutOfGas) => "ERR_OUT_OF_GAS",
+            EvmError(ExitError::OutOfFund) => "ERR_OUT_OF_FUND",
+            EvmError(ExitError::Other(m)) => m,
+            EvmError(_) => "", // unused misc
```
Comment (artob) [inline src/engine.rs:63]: Is this reachable?

File: src/engine.rs [hunk 4]
```diff
@@ -1,14 +1,100 @@
...
+            EvmFatal(ExitFatal::NotSupported) => "ERR_NOT_SUPPORTED",
+            EvmFatal(ExitFatal::UnhandledInterrupt) => "ERR_UNHANDLED_INTERRUPT",
+            EvmFatal(ExitFatal::Other(m)) => m,
+            EvmFatal(_) => "", // unused misc
```
Comment (artob) [inline src/engine.rs:67]: Is this reachable?

**Review by artob — CHANGES_REQUESTED** [review]
LGTM generally :+1: Please address the two nits and clarify reachability of the two other comments.

File: src/engine.rs (hunk 3 above)
Comment (joshuajbouw) [reply src/engine.rs:63]: Yes, there are some errors as part of `EvmError` that are flagged as unused. Not sure why they left them in as its totally dead code and nothing handles it.

File: src/engine.rs (hunk 4 above)
Comment (joshuajbouw) [reply src/engine.rs:67]: Same as the above response but for Fatal.

**Comment by artob** [comment]
> As far as I'm aware, Reverts shouldn't be considered an "ERR" as they are part of everyday EVM life? As in, the contract executes correctly, but due to some circumstance, it'll back out without actually applying changes.

Yes, EVM reverts aren't an Aurora Engine error condition, just another ordinary result from our point of view. But that needs us to extend the result types for `call`, `view`, as we already did for `submit` (nee `raw_call`). So, better handled in #48, and the way this PR handles revert is fine for the scope here.

File: src/engine.rs (hunk 3 above)
Comment (artob) [reply src/engine.rs:63]: Can we put an `unreachable` guard here instead of an empty string?

Comment (joshuajbouw) [reply src/engine.rs:85]: Sure.

**Review by artob — APPROVED** [review]
LGTM :+1: Nice work!

**Pull request merged**